File: src/main/java/fr/epicanard/mapsaver/config/Storage.java
package fr.epicanard.mapsaver.config;

import java.util.Properties;

public class Storage {
    public String Type = "mysql";
    public String TablePrefix = "MapSaver_";
    public Connection Connection;

    public String buildJdbcUrl() {
        return String.format("jdbc:%s://%s:%s/%s", this.Type, this.Connection.Host, this.Connection.Port, this.Connection.Database);
    }

    public Properties getProperties() {
        final Properties dataSourceProperties = new Properties();

        dataSourceProperties.put("useSSL", this.Connection.UseSSL);
        dataSourceProperties.put("autoReconnect", "true");
        dataSourceProperties.put("useUnicode", "true");
        dataSourceProperties.put("characterEncoding", "UTF-8");

        final Properties properties = new Properties();

        properties.put("dataSourceProperties", dataSourceProperties);
        properties.put("jdbcUrl", this.buildJdbcUrl());
        properties.put("username", this.Connection.User);
        properties.put("password", this.Connection.Password);

        return properties;
    }
}

File: src/main/scala/fr/epicanard/mapsaver/models/config/Storage.scala
package fr.epicanard.mapsaver.models.config

import io.circe.generic.JsonCodec

@JsonCodec
case class Storage(
    `type`: DatabaseType,
    tablePrefix: String,
    connection: Connection
)

object Storage {
  def toProperties(storage: Storage): Map[String, String] = Map(
    "user"              -> storage.connection.user,
    "password"          -> storage.connection.password,
    "useSSL"            -> storage.connection.useSSL.toString().toLowerCase(),
    "driver"            -> storage.`type`.driver,
    "useUnicode"        -> "true",
    "characterEncoding" -> "UTF-8",
    "autoReconnect"     -> "true"
  )

  def buildUrl(storage: Storage): String =
    s"${storage.`type`.jdbcHeader}://${storage.connection.host}:${storage.connection.port}/${storage.connection.database}"
}
